Unify and collectionify read group handling.

jmchilton opened the pull request:
The reusable macro file has:

 - Defines widgets for common read group parameters.
 - Widgets can be used at top-level (e.g. ``picard_AddOrReplaceReadGroups`` where handling is required) of inputs or within a conditional (e.g. in the mappers).
 - The conditional (used by mappers) has fours modes:
    - Set according the specification (only ID required - has all parameters). Current BWA approach.
    - Set Picard-style (presumably useful for broad pipelines?) (``ID``, ``SM``, ``LB``, ``PU``, ``PL`` are all required and there are no params for ``KS``, ``PG``, or ``FO``). In terms of what is shown and what is required - this mirrors ``picard_AddOrReplaceReadGroups``.
    - Just assign the ``ID`` and do so automatically (based either on collection or standalone dataset information). No extra form elements are shown to the user. (Should this be the new default?)
    - Do not set read groups.
 - Utilities for conditionally including parameters and formatting them (eliminates many explicit conditionals in the mappers for instance).
 - An example "mapper" test tool which demonstrates optionally including read groups.
 - An example read group assign test tool which demonstrates unconditionally specifying macros.

In addition to unifying everything, the macro file defines enhanced ``ID``, ``SM``, and ``LB`` parameter inputs that are wrapped in a conditional. That conditional allows manual input but also allows the user to just say Galaxy should use collection information or dataset name (if input is not in a collection) to assign these parameters. This should simplify things and make simple workflows more robust, but is absolutely essential for collection-based workflows and newer tool form options that allow running a tool in parallel over many inputs.

This may look at first glance like it is really complicated - but if you focus just on the mappers for instance and treat the macro file like a library - the client interface is actually a good deal easier than doing this stuff by hand for each tool.

Updates from original version:

 - Tests pass for ``picard_AddOrReplaceReadGroups`` as well as the BWA and Bowtie tools.
 - Remove the changes from tophat - it will require more work (it is less free form and won't accept certain valid read groups arguments).
 - Deleted empty lines I had previously introduced and idented the Cheetah code (hadn't realized I could do that).

Smaller Notes:

The new approach for BWA tools excludes empty read groups instead of adding them as empty - hopefully this is fine:

```
-@RG    ID:rg1  SM:     PL:CAPILLARY
-@PG    ID:bwa  PN:bwa  VN:0.7.12-r1039 CL:bwa sampe -r @RG     ID:rg1  SM:     PL:CAPILLARY localref.fa first.sai second.sai /tmp/tmpn_I5mQ/files/000/dataset_9.dat /tmp/tmpn_I5mQ/files/000/dataset_10.dat
+@RG    ID:rg1  PL:CAPILLARY
+@PG    ID:bwa  PN:bwa  VN:0.7.12-r1039 CL:bwa sampe -r @RG     ID:rg1  PL:CAPILLARY localref.fa first.sai second.sai /tmp/tmpmN3j1j/files/000/dataset_9.dat /tmp/tmpmN3j1j/files/000/dataset_10.dat
```

bgruening commented on macros/read_group_macros.xml:
```diff
@@ -0,0 +1,278 @@
+#if hasattr($input1, 'element_identifier')
+#return $input1.element_identifier
+#else
+#return $input1.name.rstrip('.gz').rstrip('.fastq').rstrip('.fq')
```
@jmchilton can we use indentation for cheetah code? This would make it so much easier to read.

nsoranzo commented on tools/bwa/bwa.xml:
```diff
@@ -109,10 +110,22 @@
 ####### Fastq paired
 
+
+
```
Don't add these empty lines?

bgruening commented:
Is `tools-devteam` the correct place for this macro? Can we include this in Galaxy so every tool can take advantage? Maybe a separate repository so we can use git submodule to manage it? I could imagine to have some additional macros that are used everywhere in such a repository, like the famous `<stdio>` macro.

erasche commented:
I'd add genome selection ("Use locally cached genome or fasta from history") to macros that I'd like to see (since that has a cheetah component and an XML component that are regularly repeated)

nekrut commented:
@erasche +1. Genome selection is very inconsistent across tools

jmchilton commented:
I am still against global macros - these make backward compatibility a nightmare.  It is better to think of them as libraries associated with your tools - Galaxy can support many more versions of tools simultaneously if the macro is defined with the tool instead of with Galaxy. Also it works better going backwards in time also - the same tool can be used in more older versions of tools.

But I have opened galaxyproject/galaxy#117 that I think address the ``<stdio>`` macro need in a much more succinct way - any objections?

I'll think more about index handling and I will address the comments from @bgruening and @nsoranzo (thanks guys).

bgruening commented:
@jmchilton what about a separate repository? I'm just thinking of a good way to distribute them to tool developers without copy and paste.

jmchilton commented:
@bgruening That could be a nice compromise - I have no problem making the ``macros`` directory I added here for instance a new repository that ``tool-devteam``, ``tools-iuc``, ``tools-other``, etc... can use as git [submodules](https://chrisjean.com/git-submodules-adding-using-removing-and-updating/).

 - Does galaxytools or tools-iuc have any tools that could use the read group stuff? Want to comment on #92 if yes?
 - Are there other macros post - galaxyproject/galaxy#117. I feel like saying it would be nice if there were a macro for genome selection is very different than there actually being such a macro (macros are kind of limited right - I am not sure they are the best approach for that - you cannot replace attributes in XML tags right - which you really would need?).

bgruening commented:
@jmchilton I have a wrapper called `segemehl` in `galaxytools` this would be one candidate. I will look over your PR, asap - still in proteomics land.

I need to think more about `macros` but my initial feeling is they are many more patterns we are frequently using, so this would make sense to share. What about the common `advanced settings` conditional? I can also imagine to have a common set of frequently used `filters` and `sanatizers` they should be used more and this would be one way to make it easier. Will think about it more.

jmchilton commented:
Okay - I updated the pull request to take into account suggestions @nsoranzo and @bgruening, fixed the tools, added a new second test case for bowtie2 (covering read group handling).

I'll rework the repository to use submodules after we have nailed this down and merged this.

bgruening commented on macros/read_group_macros.xml:
```diff
@@ -0,0 +1,294 @@
+
+    #if $rg_param('read_group_lb_conditional') is None
+        #set $rg_lb = ''
+    #elif $rg_param('read_group_lb_conditional')do_auto_name
```
`do_auto_name` looks like a left over from a comment.

bgruening commented on macros/test/read_group_mapper_1.xml:
```diff
@@ -0,0 +1,127 @@
+@set_use_rg_var@
+@set_read_group_vars@
+#if $use_rg
+echo $format_read_group("ID:", $rg_id, "'")
```
Can we use proper indentation here? This is hard to read.

bgruening commented on tools/bwa/bwa_macros.xml:
```diff
@@ -1,34 +1,18 @@
 <macros>
-
+  <import>read_group_macros.xml</import>
   <token name="@set_rg_string@">
```
@jmchilton can this megered with `set_use_rg_var`?

bgruening commented on macros/read_group_macros.xml:
```diff
@@ -0,0 +1,294 @@
+    #end if
+#end def
+
+#set $use_rg = True
```
I tried to follow the route of `$use_rg` but it's kind of complicated.
From a gut feeling I don't think we should set this to default here, rather reusing `@set_use_rg_var@`

bgruening commented on tools/bowtie2/bowtie2_wrapper.xml:
```diff
@@ -70,12 +73,30 @@
+        @define_read_group_helpers@
+        #if str( $library.type ) == "single":
+        #set $rg_auto_name = $read_group_name_default($library.input_1)
+        #elif str( $library.type ) == "paired":
```
Indentation is missing.

bgruening commented on macros/read_group_macros.xml:
```diff
@@ -0,0 +1,294 @@
+    #return $name_clean
+#end def
+
+#def read_group_name_default($input1, $input2=None)
```
If we import itertools in a function we can also import `re` here.

bgruening commented on macros/read_group_macros.xml:
```diff
@@ -0,0 +1,294 @@
+    </xml>
+    <xml name="read_group_id_param">
+        <param name="ID" type="text" value="" size="20" label="Read group identifier (ID)" help="This value must be unique among multiple samples in your experiment" optional="false">
+        <validator type="empty_field" />
```
indentation

bgruening commented:
This is a very nice work! Thanks @jmchilton! 
I would appreciate more reviews as this will change a lot of tools. @nsoranzo @blankenberg please :)

John, it would be nice if you could add a few comments to your macros or maybe a small readme file how to use these properly. 

Thanks again!
Bjoern

bgruening commented:
@nsoranzo @blankenberg @nekrut @peterjc this has a huge impact more eyes please!

peterjc commented:
The fields etc all look SAM/BAM format related, which makes sense. Is the expectation that these macros ought to be useful for any mapping tool outputting SAM/BAM files?

This isn't a criticism, just an observation: It looks like because MIRA4 does things a little differently (e.g. it has a similar but different table of sequencing technology names), I would not be able to use many of the macros there: https://github.com/peterjc/pico_galaxy/tree/master/tools/mira4_assembler

jmchilton commented:
@bgruening IMO there is no need to rush this PR - lets wait at least for Anton's approval before merging it. I'll try to find sometime to run into the office and demo it for him. 

@peterjc Yeah - they are definitely not going to work for everything. Tophat doesn't allow setting all the fields available in the SAM/BAM spec for instance - I've considered looking into a pull request to fix that.

nekrut commented:
I think this is beautiful! Merge!!!!!!!!!!!!

bgruening commented:
If the boss is so enthusiastic about this I would like to merge this. 
@martenson I will merge this as soon as BWA was updated. It's probably a good idea to not push this directly to the MTS.

jmchilton commented:
So you don't trust me with the MTS right away huh? Wise - very wise.

bgruening commented:
:( sorry @jmchilton. It's just such a large change ...

martenson commented:
@bgruening I do not understand this sentence:  "I will merge this as soon as BWA was updated."

bgruening commented:
@martenson can you please update BWA to fix https://github.com/galaxyproject/tools-devteam/issues/157 for Jen

martenson commented:
@bgruening done (https://toolshed.g2.bx.psu.edu/view/devteam/bwa/ac30bfd3e2a8) thank for the ping!

bgruening commented:
Awesome! Thanks a lot!

bgruening commented:
Thanks everyone! This is a big great step!
